### PR TITLE
WT-10108 Add a data structure encapsulating user level truncate context

### PR DESCRIPTION
**Pull request by agorrod**

### CODE REVIEW

File: src/include/truncate.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    WT_CURSOR *start;
+    WT_CURSOR *stop;
+    WT_ITEM *orig_start_key;
+    WT_ITEM *orig_stop_key;
+
+/* AUTOMATIC FLAG VALUE GENERATION START 0 */
+#define WT_TRUNC_EXPLICIT_START 0x1u
+#define WT_TRUNC_EXPLICIT_STOP 0x2u
+    /* AUTOMATIC FLAG VALUE GENERATION STOP 32 */
+    uint32_t flags;
```
Comment (quchenhao) [inline src/include/truncate.h:25]: Is 8 bit enough?

File: src/cursor/cur_hs.c [hunk 2]
```diff
@@ -1181,20 +1181,22 @@ __curhs_update(WT_CURSOR *cursor)
...
 {
     WT_CURSOR *start_file_cursor, *stop_file_cursor;
     WT_SESSION_IMPL *session;
 
-    session = CUR2S(start);
-    start_file_cursor = ((WT_CURSOR_HS *)start)->file_cursor;
-    stop_file_cursor = stop != NULL ? ((WT_CURSOR_HS *)stop)->file_cursor : NULL;
+    session = trunc_info->session;
+    stop_file_cursor = NULL;
+
```
Comment (luke-pearson) [inline src/cursor/cur_hs.c:1191]: This whitespace seems unnecessary

File: src/include/truncate.h (hunk 1 above)
Comment (agorrod) [reply src/include/truncate.h:25]: It is, but I tend to default to 32 bit values unless there is a reason to be space sensitive, which I don't think there is in this case.
If you feel strongly, I can change it.

Comment (quchenhao) [reply src/include/truncate.h:25]: No, I don't.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/cursor/cur_hs.c [hunk 3]
```diff
@@ -1181,20 +1181,22 @@ __curhs_update(WT_CURSOR *cursor)
...
+    start_file_cursor = ((WT_CURSOR_HS *)trunc_info->start)->file_cursor;
 
     WT_STAT_DATA_INCR(session, cursor_truncate);
 
     WT_ASSERT(session, F_ISSET(start_file_cursor, WT_CURSTD_KEY_INT));
     WT_RET(__wt_cursor_localkey(start_file_cursor));
-    if (stop != NULL) {
+    if (F_ISSET(trunc_info, WT_TRUNC_EXPLICIT_STOP)) {
+        stop_file_cursor = ((WT_CURSOR_HS *)trunc_info->stop)->file_cursor;
```
Comment (luke-pearson) [inline src/cursor/cur_hs.c:1199]: I feel like this HS_CURSOR cast could use its own macro one day but not required for this PR.

File: src/txn/txn_log.c [hunk 4]
```diff
@@ -594,28 +595,40 @@ __wt_txn_truncate_log(
          * If the user provided a start cursor key (i.e. local_start is false) then use the original
          * key provided.
          */
-        if (!local_start && orig_start_key != NULL) {
+        if (F_ISSET(trunc_info, WT_TRUNC_EXPLICIT_START)) {
+            WT_ASSERT_ALWAYS(session, trunc_info->orig_start_key != NULL,
+              "Logging truncate operation with explicit start to range and empty key");
```
Comment (luke-pearson) [inline src/txn/txn_log.c:600]: I think this assertion message is confusing, I would prefer something like "Attempted to log truncate operation with explicit start to range but empty key".

File: src/cursor/cur_hs.c (hunk 2 above)
Comment (sueloverso) [reply src/cursor/cur_hs.c:1191]: Also alphabetize the lines to set start first.

File: src/cursor/cur_table.c [hunk 5]
```diff
@@ -748,17 +748,20 @@ __curtable_reserve(WT_CURSOR *cursor)
  *     Truncate of a cursor range, table implementation.
  */
 int
-__wt_table_range_truncate(WT_CURSOR_TABLE *start, WT_CURSOR_TABLE *stop)
+__wt_table_range_truncate(WT_TRUNCATE_INFO *trunc_info)
 {
     WT_CURSOR *wt_start, *wt_stop;
     WT_CURSOR_TABLE *ctable;
+    WT_CURSOR_TABLE *start, *stop;
```
Comment (sueloverso) [inline src/cursor/cur_table.c:755]: This should be merged with the `ctable` line above it.

File: src/cursor/cur_table.c [hunk 6]
```diff
@@ -810,9 +813,10 @@ __wt_table_range_truncate(WT_CURSOR_TABLE *start, WT_CURSOR_TABLE *stop)
     }
 
     /* Truncate the column groups. */
-    for (i = 0; i < WT_COLGROUPS(ctable->table); i++)
+    for (i = 0; i < WT_COLGROUPS(ctable->table); i++) {
```
Comment (sueloverso) [inline src/cursor/cur_table.c:816]: This change is not necessary. Were the braces added for debugging statements and then left in?

File: src/session/session_api.c [hunk 7]
```diff
@@ -1487,13 +1487,24 @@ __wt_session_range_truncate(
...
+    trunc_info = &_trunc_info;
+    memset(trunc_info, 0, sizeof(*trunc_info));
+    if (uri == NULL && start != NULL)
+        F_SET(trunc_info, WT_TRUNC_EXPLICIT_START);
+    if (uri == NULL && stop != NULL)
+        F_SET(trunc_info, WT_TRUNC_EXPLICIT_STOP);
+
     if (uri != NULL) {
         WT_ASSERT(session, WT_BTREE_PREFIX(uri));
+
```
Comment (sueloverso) [inline src/session/session_api.c:1507]: Remove whitespace.

File: src/txn/txn_log.c (hunk 4 above)
Comment (sueloverso) [reply src/txn/txn_log.c:600]: I agree it's confusing. My suggestion would be something like "Truncate log operation with explicit range start has empty original key."

If you choose to change this message, all the others should be similarly changed.

**Review by sueloverso — APPROVED** [review]
LGTM

**Pull request merged**